[WIP] Update `past_key_values` in GPT-2

forest1988 opened the pull request:
# What does this PR do?

It seems GPT-2 and BartDecoder has a different style of `past_key_values`.
Advised by @patrickvonplaten,
I opened this PR to change GPT-2's cache format from a single tensor to a tuple of 2 tensors.
Once this problem is solved, it is expected that `past_key_values` in GPT-2 will be handled in the same way as in Bart.

Sorry there remain some errors. This PR is [WIP].
I would appreciate your advice on how to update `generation_utils.py`.
Can I modify `_reorder_cache` so that past is replaced from Tuple[torch.Tensor] to Tuple[Tuple[torch.Tensor]],
or should I consider other output variations, output.mem and outputs.past_buckets_states?

Fixes #9391


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

GPT2: @LysandreJik, @patrickvonplaten

forest1988 commented:
CircleCI error messages says as below.

In `run_tests_torch`:
``` 
-- Docs: https://docs.pytest.org/en/stable/warnings.html
=========================== short test summary info ============================
FAILED tests/test_modeling_gpt2.py::GPT2ModelTest::test_beam_sample_generate
FAILED tests/test_modeling_gpt2.py::GPT2ModelTest::test_beam_search_generate
FAILED tests/test_modeling_gpt2.py::GPT2ModelTest::test_beam_search_generate_dict_outputs_use_cache
FAILED tests/test_modeling_gpt2.py::GPT2ModelTest::test_gpt2_gradient_checkpointing
FAILED tests/test_modeling_gpt2.py::GPT2ModelTest::test_group_beam_search_generate
==== 5 failed, 4202 passed, 1775 skipped, 744 warnings in 216.47s (0:03:36) ====

Exited with code exit status 1
CircleCI received exit code 1
```

In `run_tests_flax`:
```
FAILED tests/test_modeling_gpt2.py::GPT2ModelTest::test_beam_sample_generate
FAILED tests/test_modeling_gpt2.py::GPT2ModelTest::test_beam_search_generate
FAILED tests/test_modeling_gpt2.py::GPT2ModelTest::test_beam_search_generate_dict_outputs_use_cache
FAILED tests/test_modeling_gpt2.py::GPT2ModelTest::test_gpt2_gradient_checkpointing
FAILED tests/test_modeling_gpt2.py::GPT2ModelTest::test_group_beam_search_generate
==== 5 failed, 4172 passed, 1805 skipped, 751 warnings in 282.27s (0:04:42) ====

Exited with code exit status 1
CircleCI received exit code 1
```

forest1988 commented:
Is there a difference between `past_key_value` and `layer_past`?　I understand that they both represent the contents of `past_key_values`, the past of each layer, but are they different?

I first thought it might be a difference between the Causal language model and the Seq2Seq language model, but it seems that both `past_key_value` and `layer_past` are used in `modeling_bart.py`.

And as for the contents of `layer_past`, should it be named `past_state`, as the following part of `modeling_bart.py` shows?

https://github.com/huggingface/transformers/blob/236cc365aff2512ef773c6b1786555dab6fb182f/src/transformers/models/bart/modeling_bart.py#L1236-L1244

forest1988 commented:
I've updated `generation_utils.py`, and it seems `mems` in transfo_xl and xlnet causes a new error.

```
=========================== short test summary info ============================
FAILED tests/test_modeling_gpt2.py::GPT2ModelTest::test_gpt2_gradient_checkpointing
FAILED tests/test_modeling_transfo_xl.py::TransfoXLModelTest::test_beam_sample_generate
FAILED tests/test_modeling_transfo_xl.py::TransfoXLModelTest::test_beam_sample_generate_dict_output
FAILED tests/test_modeling_transfo_xl.py::TransfoXLModelTest::test_beam_search_generate
FAILED tests/test_modeling_transfo_xl.py::TransfoXLModelTest::test_beam_search_generate_dict_output
FAILED tests/test_modeling_transfo_xl.py::TransfoXLModelTest::test_group_beam_search_generate
FAILED tests/test_modeling_transfo_xl.py::TransfoXLModelTest::test_group_beam_search_generate_dict_output
FAILED tests/test_modeling_xlnet.py::XLNetModelTest::test_beam_sample_generate
FAILED tests/test_modeling_xlnet.py::XLNetModelTest::test_beam_sample_generate_dict_output
FAILED tests/test_modeling_xlnet.py::XLNetModelTest::test_beam_search_generate
FAILED tests/test_modeling_xlnet.py::XLNetModelTest::test_beam_search_generate_dict_output
FAILED tests/test_modeling_xlnet.py::XLNetModelTest::test_group_beam_search_generate
FAILED tests/test_modeling_xlnet.py::XLNetModelTest::test_group_beam_search_generate_dict_output
=== 13 failed, 4194 passed, 1775 skipped, 743 warnings in 205.38s (0:03:25) ====

Exited with code exit status 1
CircleCI received exit code 1
```

https://github.com/huggingface/transformers/blob/236cc365aff2512ef773c6b1786555dab6fb182f/src/transformers/models/xlnet/modeling_xlnet.py#L581-L607

It seems `mems` is something similar to `past_key_values`. 
Is there any difference between these two elements with different names?
Also, is it safe to change `mems` from `List[torch.Tensor]` to `Tuple[Tuple[torch.Tensor]]`?

patrickvonplaten commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -232,7 +232,7 @@ def forward(
 
         if use_cache is True:
-            present = torch.stack((key.transpose(-2, -1), value))  # transpose to have same shapes for stacking
+            present = (key.transpose(-2, -1), value)  # transpose to have same shapes
```
this looks nice!

patrickvonplaten replied:
This is the reason for the recent failure of the slow test:

`RUN_SLOW=1 pytest tests/test_onnx.py::OnnxExportTestCase::test_export_pytorch` 

Can you fix the onnx part easily? @mfuntowicz @Narsil

patrickvonplaten commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -369,9 +369,9 @@ class GPT2DoubleHeadsModelOutput(ModelOutput):
-        past_key_values (:obj:`List[torch.FloatTensor]`, `optional`, returned when ``use_cache=True`` is passed or when ``config.use_cache=True``):
-            List of :obj:`torch.FloatTensor` of length :obj:`config.n_layers`, with each tensor of shape :obj:`(2,
-            batch_size, num_heads, sequence_length, embed_size_per_head)`).
+        past_key_values (:obj:`Tuple[Tuple[torch.Tensor]]`, `optional`, returned when ``use_cache=True`` is passed or when ``config.use_cache=True``):
```
nice!

patrickvonplaten commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -707,7 +707,7 @@ def forward(
                 # Ensure layer_past is on same device as hidden_states (might not be correct)
                 if layer_past is not None:
-                    layer_past = layer_past.to(hidden_states.device)
+                    layer_past = tuple(past_state.to(hidden_states.device) for past_state in layer_past)
```
nice!

patrickvonplaten commented:
Hey @forest1988, 

You're PR looks very nice! Yes, it is expected that `XLNet` and `TransfoXL` fail actually since they also have been using the "default" `_reorder_cache` function of `modeling_utils.py`. Could you do the following changes to correct this:

1) Copy that old `_reorder_cache` (the one before you did your changes) function that was in `generation_utils.py` to both `modeling_xlnet.py` and `modeling_transfo_xl.py` file so that those have the same function as before? 
2) Copy the current `_reorder_cache` function of `generation_utils.py` into `modeling_gpt2.py`?
3) Add a default `_reorder_cache` function to `generation_utils.py` that looks as follows:

```python
def _reorder_cache(self, past, beam_idx):
    raise NotImplementedError(...)
```

forest1988 commented:
I've just updated `torch.utils.checkpoint.checkpoint` check in `modeling_gpt2.py`, referring to `modeling_bart.py`.

patrickvonplaten commented:
This way it's much cleaner and correct :-) The reason I'm proposing this change is that the `_reorder_cache` function is so different for each model that there should be **no** default function. A default function could confuse people that want to add a new model in a way that they think it works out of the box, but in most cases it just doesn't. A clear error message such as:


```python
def _reorder_cache(self, past, beam_idx):
    raise NotImplementedError(f"Make sure that a `_reorder_cache` function is correctly implemented in {self.__class__.__module__} to enable beam search for {self.__class__}")
```
```

patrickvonplaten commented:
I think this should solve the problems, let me know if you need more help :-)

forest1988 commented:
Thank you for your advice! I'll update `_reorder_cache` soon and commit it.

forest1988 commented:
Hi @patrickvonplaten,

Thanks to your kind advice, I could solve the problem of `_reorder_cache` in GPT-2, XLNet, TransfoXL (, and CTRL).
Referring to `modeling_bart.py`, in which `_reorder_cache` is placed in `ConditionalGeneration` Model, I added `_reoder_cache` in `LMHead` Models in each Causal Language Models.

The last one remaining bug is:
```
FAILED tests/test_modeling_gpt2.py::GPT2ModelTest::test_gpt2_gradient_checkpointing
```

I think I should modify `test_gpt2_gradient_checkpointing` so that it has `use_cache=False`, or reconsider my previous update and re-modify the usage of `checkpoint` in modeling_gpt2.
> I've just updated torch.utils.checkpoint.checkpoint check in modeling_gpt2.py, referring to modeling_bart.py.
>

forest1988 commented:
All checks have passed!
I appreciate all your help.

However, in the documentation of `_reorder_cache`, there are references to both `past_key_values` and `mems` regardless of which object is used.
I think we can fix that and only mention the one we use, or we can leave the reference to both to show that the aim of the function is the same.
If there is a need to modify it, please let me know.

patrickvonplaten commented on src/transformers/models/ctrl/modeling_ctrl.py:
```diff
@@ -575,6 +577,23 @@ def forward(
+        called. This is required to match :obj:`past_key_values` or :obj:`mems` with the correct beam_idx at every
+        generation step.
+
+        For custom re-ordering of :obj:`past_key_values` or :obj:`mems`, the function should be implemented in
```
Yeah it would be nice if you could in a final step clean up the doc string here removing those two lines and removing all  mentioning of mems above

forest1988 replied:
Thanks! Referring to your comments, I've just done the clean up of the doc strings.

patrickvonplaten commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -717,18 +717,22 @@ def forward(
                 all_hidden_states = all_hidden_states + (hidden_states,)
 
             if getattr(self.config, "gradient_checkpointing", False):
+                if use_cache:
```
I agree with this change, I don't see how and why `use_cache` could be used with `gradient_checkpointing`. The only problem I can see if someone trains GPT2 with gradient checkpointing, but wants to enable `use_cache` for validation and evaluation afterward. IMO, we should however force `use_cache=False` during train, and `gradient_checkpointing=False` during validation  in the `Trainer`? Let's see what @LysandreJik and @sgugger think

sgugger replied:
Why would `gradient_checkpointing` be `True` during validation? I'm a bit lost.

patrickvonplaten replied:
Explained offline

patrickvonplaten commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -931,6 +935,23 @@ def forward(
+    @staticmethod
+    def _reorder_cache(past: Tuple[Tuple[torch.Tensor]], beam_idx: torch.Tensor) -> Tuple[Tuple[torch.Tensor]]:
+        """
+        This function is used to re-order the :obj:`past_key_values` or :obj:`mems` cache if
```
Same here, it would be nice if we can clean up the docs

forest1988 replied:
I cleaned it as well.

patrickvonplaten commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -1094,6 +1115,23 @@ def forward(
+    @staticmethod
+    def _reorder_cache(past: Tuple[Tuple[torch.Tensor]], beam_idx: torch.Tensor) -> Tuple[Tuple[torch.Tensor]]:
+        """
+        This function is used to re-order the :obj:`past_key_values` or :obj:`mems` cache if
```
also remove `mems` here if possible

forest1988 replied:
I cleaned it as well.

patrickvonplaten commented on src/transformers/models/transfo_xl/modeling_transfo_xl.py:
```diff
@@ -1137,6 +1137,19 @@ def _resize_cutoffs(self, new_num_tokens, new_emb_size, new_embedding_shapes, la
+        called. This is required to match :obj:`past_key_values` or :obj:`mems` with the correct beam_idx at every
+        generation step.
+
+        For custom re-ordering of :obj:`past_key_values` or :obj:`mems`, the function should be implemented in
```
remove those lines and `past_key_values` above

forest1988 replied:
I cleaned it as well.

patrickvonplaten commented on src/transformers/models/xlnet/modeling_xlnet.py:
```diff
@@ -1462,6 +1462,19 @@ def forward(
+        called. This is required to match :obj:`past_key_values` or :obj:`mems` with the correct beam_idx at every
+        generation step.
+
+        For custom re-ordering of :obj:`past_key_values` or :obj:`mems`, the function should be implemented in
```
remove those lines and `past_key_values` above

forest1988 replied:
I cleaned it as well.

patrickvonplaten reviewed (commented):
The PR looks very nice - thanks so much for taking the time to tackle this @forest1988 . Let's wait a bit to see how to proceed with `gradient_checkpointing` in GPT2 as this question will come up more often. IMO, `use_cache` should always be `False` for training so either we update all `use_cache` in the models with a `use_cache= not self.is_training and (use_cache if use_cache is not None else self.config.use_cache)` or we force it somehow in the Trainer. Similarly `gradient_checkpointing` should never be set to True when the model is not training IMO (we could also automatically disable this using `self.training`). Let's see what @LysandreJik and @sgugger think.

sgugger commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -369,9 +369,9 @@ class GPT2DoubleHeadsModelOutput(ModelOutput):
-            List of :obj:`torch.FloatTensor` of length :obj:`config.n_layers`, with each tensor of shape :obj:`(2,
-            batch_size, num_heads, sequence_length, embed_size_per_head)`).
+        past_key_values (:obj:`Tuple[Tuple[torch.Tensor]]`, `optional`, returned when ``use_cache=True`` is passed or when ``config.use_cache=True``):
+            :obj:`Tuple[Tuple[torch.Tensor]]` of length :obj:`config.n_layers`, with each tuple having 2 tuples each of
+            which has a tensor of shape :obj:`(batch_size, num_heads, sequence_length, embed_size_per_head)`).
```
Let's use plain English in the description, the type is already written above.
```suggestion
           Tuple of length :obj:`config.n_layers`, containing tuples of tensors of shape
            :obj:`(batch_size, num_heads, sequence_length, embed_size_per_head)`).
```

sgugger reviewed (approved):
This is not a part of the library I'm very familiar with, so the changes look okay on my side, but I'm no expert.

LysandreJik commented on tests/test_modeling_gpt2.py:
```diff
@@ -131,6 +131,7 @@ def prepare_config_and_inputs(self, gradient_checkpointing=False):
             n_ctx=self.max_position_embeddings,
             # type_vocab_size=self.type_vocab_size,
             # initializer_range=self.initializer_range,
+            use_cache=not gradient_checkpointing,
```
Cool!

LysandreJik reviewed (approved):
These changes look good to me! Thanks for taking care of it @forest1988.

patrickvonplaten reviewed (approved):
Great work @forest1988,

I hope it's fine for you that I went into the PR to do some final fixes. Thanks a lot for cleaning this up :-)

forest1988 commented:
Hi @patrickvonplaten,

> I hope it's fine for you that I went into the PR to do some final fixes. Thanks a lot for cleaning this up :-)

Of course! Thank you for adding fixes to make this PR more valuable!

LysandreJik reviewed (approved):
Your commit looks good to me @patrickvonplaten! Thanks.

sgugger reviewed (approved):
The new changes look good to me, thanks!

patrickvonplaten commented:
Awesome, merging - great job @forest1988 !

forest1988 commented:
Thank you for your advice and encouraging comments!
It’s my pleasure to have opened this PR!
